File: Python/15-3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        
        res = []
        n = len(nums)
        
        # sort array to avoid redundant solutions
        nums.sort()
        
        for i, curr_num in enumerate(nums):
            
            # skip equal numbers to avoid redundant solutions
            if i > 0 and curr_num == nums[i - 1]:
                continue
            
            # create left and right pointer to run through remaining nums
            l,r = i + 1, n - 1
            while l < r:
                
                curr_sum = curr_num + nums[l] + nums[r]
                
                if curr_sum < 0:
                    l += 1
                
                elif curr_sum > 0:
                    r -= 1
                    
                else: # == 0
                    res.append([curr_num, nums[l], nums[r]])
                    l += 1
                    
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1
                    
        return res

File: Java/15-3Sum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
    	
    	List<List<Integer>> res = new ArrayList<List<Integer>>();
    	int n = nums.length;
    	
    	// sort array to avoid redundant solutions
    	Arrays.sort(nums);
    	
    	for(int i = 0; i < n; i++) {
    		
    		int curr_num = nums[i];
    		
    		if(i > 0 && curr_num == nums[i-1]) continue;
    		
    		// left and right pointer to find three sum
    		int l = i + 1;
    		int r = n - 1;
    		
    		while(l < r) {
    			
    			int curr_sum = curr_num + nums[l] + nums[r];
    			
    			// move pointer 
    			l = (curr_sum < 0)? l+1 : l;
    			r = (curr_sum > 0)? r-1 : r;
    			
    			// three sum found
    			if(curr_sum == 0) {
    				res.add(Arrays.asList(curr_num, nums[l], nums[r]));
    				l += 1;
    				
    				while(nums[l] == nums[l - 1] && l < r) l+=1 ;
    			}
    		}
    	}
    	
		return res;
    }
}